### DISCUSSION TITLE
Large databases fail to load in admin UI

### DISCUSSION THREAD

**Discussion by jonshipman** [pocketbase/pocketbase#3859] [Q&A] [answered]
Database with 146764 records does not load in the admin ui. Works fine via $app.dao calls and the web rest api, but not in the admin. data.db is 6.8G and Amazon S3 backups are enabled.

`{"code":400,"message":"Invalid filter parameters.","data":{}}`

**Comment by ganigeorgiev** [comment]
The database size shouldn't matter as long as the execution time is below the query timeout (by default is ~30sec).

I have a local database with 1m+ records so something else is going on in your case but without reproducible steps or at least some details about your collections, I don't know how to help.

What PocketBase version do you use?

Are you providing a filter expression when the above error is returned?

What is the type of collection(s) you are trying to view in the Admin UI (base, auth, view)?

Could you paste your collections as json (Admin UI > Settings > Export collections)?

**Reply by ganigeorgiev** [reply]
If you've created the collection tables manually or from some import script then I guess it is possible something to have messed up with the schema definitions, but again without details and some context I don't know how to reproduce it.

**Reply by jonshipman** [reply]
[pb_schema.json](https://github.com/pocketbase/pocketbase/files/13538861/pb_schema.json)

**Reply by jonshipman** [reply]
I'll also see if I can get you a pb_data folder downloadable

**Reply by ganigeorgiev** [reply]
@jonshipman I'm still not able to reproduce the reported issue.

Are you even sure that "Invalid filter parameters" is from the Admin UI failed request? How did you get that log?

I've tried to insert 200k records in the `weather_alert` collection and it load just fine for me.

If you need further help, please:

- report what version of PocketBase do you use
- if the issue you are experiencing is only in the Admin UI, then please check if there is a JS error in the browser console
- start the PocketBase executable with `--debug` and check if there any more detailed logs printed in the stder
- provide a code sample of your custom JS hooks (if any)

**Reply by ganigeorgiev** [reply]
Also note that if you have custom Go or JS hooks code, this could also affect the Admin UI requests because the Admin UI uses the same web APIs.

**Reply by jonshipman** [reply]
This should reproduce the error.
https://issue-3859.jonshipman.com/3859.tar.gz

jon@admin.com
qkS4A108nylAqjCIBpoPVXnU0BHD4lAv

**Reply by jonshipman** [reply]
v0.19.4

**Reply by jonshipman** [reply]
<img width="948" alt="image" src="https://github.com/pocketbase/pocketbase/assets/12699954/24c8f11d-fc94-4bbf-9b2d-6cb12628d4e7">

**Reply by jonshipman** [reply]
output from --debug

[30072.00ms] SELECT `weather_alert`.* FROM `weather_alert` ORDER BY [[weather_alert.created]] DESC LIMIT 40
2023/12/03 07:32:41 interrupted (9)

**Reply by ganigeorgiev** [reply, marked as answer]
It is the query timeout. Try to increase it for example to 60s:
```
./pocketbase serve --queryTimeout=60
```

Why it fails only from the Admin UI and not locally I'm not sure - maybe your client request has a different limit or uses a filter/sort that make use of an index?

But taking more than 30s for a plain select in ~150k table is definitely strange.

Locally I'm not able to reproduce it, but my guess is that it would depend on the size of the data stored in your JSON fileds `properties` and `geometry`.

To rule out PocketBase issue, you can also try to execute the query directly against the SQL database using the [sqlite3 cli](https://sqlite.org/cli.html) tool:

```sh
sqlite3 ./pb_data/data.db

> .timer on
> SELECT weather_alert.* FROM weather_alert ORDER BY weather_alert.created DESC LIMIT 40
```
Note that by default we use a pure Go port of SQLite so the execution is expected to be slightly slower compared to the sqlite3 cli tool.

You can try to define an index on the `created` field from the Admin UI > Edit collection > "Unique constraints and indexes" section to see if that will improve a little the loading times.

To avoid downloading your entire database, could you export/paste sample data of a single `weather_alert` record and more specifically the JSON fields `properties` and `geometry`?

**Reply by jonshipman** [reply]
[1.json](https://github.com/pocketbase/pocketbase/files/13539159/1.json)

**Reply by jonshipman** [reply]
sqlite3 time

Run Time: real 37.560 user 14.251689 sys 17.127511

**Reply by jonshipman** [reply]
Here's a larger sample 
[2.json](https://github.com/pocketbase/pocketbase/files/13539198/2.json)

**Reply by ganigeorgiev** [reply]
Hm, sqlite3 also seems to take 30s+ for the query execution and I'm not sure why exactly.

I suspect the large json fields for some reason slows the query. To confirm this you can again try to run the query but without these 2 fields, something like:

```sql
SELECT id, active, created, effective, hailSize, noaaId, state, updated, windGust
FROM weather_alert
ORDER BY weather_alert.created DESC
LIMIT 40
```

I'm still not able to reproduce it by bulk inserting `1.json` 150k times (only the `noaaId` is different for each record), but I'll try to download your `pb_data` and will test it sometime later today.

**Reply by jonshipman** [reply]
I increased the timeout to  --queryTimeout=1200 and it finally loaded. Probably would work with a smaller number but it's some headway at least.

I ran your newer query wo those two fields and got `Run Time: real 21.636 user 1.097611 sys 5.468161`

**Reply by jonshipman** [reply]
I think I found the issue:

<img width="446" alt="image" src="https://github.com/pocketbase/pocketbase/assets/12699954/38c41183-9e2b-4569-b35f-47e488b69d15">

When I removed state and active from the index and had created by itself, it loaded instantaneously.

**Reply by ganigeorgiev** [reply]
I'm glad that you've found the culprit.

Indeed, it seems that the SQLite query planner is not able to take an efficient route with the combined index for this particular query. From testing locally with your `pb_data`, the size of each individual db entry also seems to affect the query execution significantly (I guess if there is no proper index it still has to copy some of the entries in memory so that it can extract the order columns and sort them).

For completeness, here are the query plans with and without the combined index:

```
-- combined state,created,active index
|--SCAN weather_alert
 --USE TEMP B-TREE FOR ORDER BY

-- individual created index
|--SCAN weather_alert USING INDEX idx_yUeBQJg
```
 
If you are still having performance issues with the queries down the road, you could:

- find the slow queries and inspect its query plan with the sqlite3 cli tool using `EXPLAIN QUERY PLAN yourQuery` and/or use the [`.expert` sqlite3 cli](https://sqlite.org/cli.html#special_commands_to_sqlite3_dot_commands_) mode to suggest the indexes (_the latter eventually may get integrated in the PocketBase Admin UI_)
- refactor your collection and split the content in several other collections to try to keep the overall individual record size "small"

There are several other minor optimizations we can perform internally in PocketBase like increasing the SQLite page size, running [`PRAGMA optimize`](https://www.sqlite.org/pragma.html#pragma_optimize) periodically, etc., but that will be done after https://github.com/pocketbase/pocketbase/discussions/3271.

**Comment by jonshipman** [comment]
Moving to a feature request
